ui/hooks/useNftCollectionsMetadata.ts

```diff
@@ -0,0 +1,112 @@
+import { useEffect, useMemo, useState } from 'react';
+import { useSelector } from 'react-redux';
+import { Hex } from '@metamask/utils';
+import { TokenStandard } from '../../shared/constants/transaction';
+import { hexToDecimal } from '../../shared/modules/conversion.utils';
+import { getCurrentChainId } from '../selectors';
+import {
+  getNFTTokenInfo,
+  getNftInformationFromTokenURI,
+} from '../store/actions';
+
+export type UseNftCollectionsMetadataRequest = {
+  standard?: TokenStandard;
+  tokenId?: Hex;
+  value: string;
+};
+
+type NFTCollection = {
+  name?: string;
+  image?: string;
+  isSpam: boolean | null;
+};
+
+type NFTCollectionData = {
+  [key: string]: NFTCollection;
+};
+
+type TokenURIResponse = {
+  name?: string;
+  image?: string;
+};
+
+const NFT_TOKEN_STANDARDS = [TokenStandard.ERC1155, TokenStandard.ERC721];
+
+export function useNftCollectionsMetadata(
+  requests: UseNftCollectionsMetadataRequest[],
+) {
+  const chainId = useSelector(getCurrentChainId);
+  const [collectionsMetadata, setCollectionsMetadata] =
+    useState<NFTCollectionData>({});
+  const nftRequests = requests
+    .filter(({ standard }) =>
+      NFT_TOKEN_STANDARDS.includes(standard as TokenStandard),
+    )
+    .filter(({ value }) => value);
+
+  const memoisedNFTRequests = useMemo(() => {
+    return nftRequests.map(({ value, tokenId, standard }) => ({
+      contractAddress: value,
+      tokenId: hexToDecimal(tokenId as string),
+      standard,
+    }));
+  }, [JSON.stringify(nftRequests)]);
+
+  useEffect(() => {
+    const fetchTokenURIData = async () => {
+      const tokenURIPromises = memoisedNFTRequests.map(
+        ({ contractAddress, tokenId }) =>
+          getNftInformationFromTokenURI(contractAddress, tokenId),
+      ) as unknown as Promise<TokenURIResponse>[];
+      const tokenURIResults = await Promise.all(tokenURIPromises);
+
+      if (tokenURIResults.some((result) => !result.name)) {
+        throw new Error('TokenURI response is missing name, skipping');
+      }
+
+      return tokenURIResults.reduce((acc, { name, image }, index) => {
+        const { contractAddress, tokenId } = memoisedNFTRequests[index];
+        acc[`${contractAddress}:${tokenId}`] = {
+          name,
+          image,
+          // We don't have a way to determine if a collection is spam from the tokenURI
+          isSpam: null,
```

Comment:
If we can't determine if a contract `isSpam`, then the data is unusable for the petnames use case.

That's the central mechanism we need to determine if it's safe to display a default petname.

Reply:
Done